virtual_rainforest/core/data.py

```diff
@@ -0,0 +1,686 @@
+        # signature already exists
+        if signature in AXIS_VALIDATORS[axis]:
+            LOGGER.debug("Replacing existing %s validator: %s", axis, func.__name__)
+        else:
+            LOGGER.debug("Adding %s validator: %s", axis, func.__name__)
+
+        # Add the validator
+        AXIS_VALIDATORS[axis][signature] = func
+
+        return func
+
+    return decorator
+
+
+def get_validator(axis: str, data: Data, darray: DataArray) -> Optional[Callable]:
+    """Get the matching validator function for a data array on a given axis.
+
+    This function iterates over the registered validator functions for a given core axis
+    in the AXIS_VALIDATORS registry and matches the data array signatures against the
+    provided signatures. It returns either the appropriate function or None.
+
+    Args:
+        axis: The core axis to get a validator for
+        data: A Data object providing parameter information
+        darray: The data to match to a validator signature
+    """
+
+    # Check the axis exists
+    if axis not in AXIS_VALIDATORS:
+        raise ValueError(f"Unknown core axis: {axis}")
+
+    # Identify the correct validator routine from the data array signature
+    da_dims = set(darray.dims)
+    da_coords = set(darray.coords.variables)
+
+    # Variable to hold the matching callable if found
+    matching_validator: Optional[Callable] = None
+
+    # Loop over the available validators loooking for a congruent signature
+    # - dims _cannot_ be empty on a DataArray (they default to dim_N strings) so are
+    #   guaranteed to be non-empty and must then match.
+    # - coords _can_ be empty: they just associate values with indices along the
+    #   dimension. So, the signature should match _specified_ coords but not the
+    #   empty set.
+    for (ld_dims, ld_coords, ld_grid_type), ld_fnm in AXIS_VALIDATORS[axis].items():
+
+        if (
+            set(ld_dims).issubset(da_dims)
+            and set(ld_coords).issubset(da_coords)
+            and ((data.grid.grid_type in ld_grid_type) or ("__any__" in ld_grid_type))
+        ):
+
+            # Retrieve the method associated with the loader signature from the Data
+            # object.
+            matching_validator = ld_fnm
+            break
+
+    return matching_validator
+
+
+@register_axis_validator("spatial", (("cell_id",), ("cell_id",), ("__any__",)))
+def spld_cellid_coord_any(data: Data, darray: DataArray) -> DataArray:
+    """Spatial loader for cell id coordinates onto any grid.
+
+    In this loader, the DataArray has a cell_id dimension with valued coordinates, which
+    should map onto the grid cell ids, allowing for a subset of ids. Because this method
+    simply maps data to grid cells by id, it should applies to _any_ arbitrary grid
+    setup.
+    """
+
+    da_cell_ids = darray["cell_id"].values
+
+    if len(np.unique(da_cell_ids)) != len(da_cell_ids):
+        raise ValueError("The data cell ids contain duplicate values.")
```

Comment:
The duplicate cell ids map onto probably different values in the data array though. This can happen if a data array has more than one value mapped to the same coordinate.

```python
meh_data = DataArray(
    np.random.normal(loc=20.0, size=(10,)),
    name="temperature",
    coords={'cell_id': [1,2,3,4,5,6,7,8,9,9]},
)
```